[BUG] 24.06 test_conditional_with_side_effects_case_when test failed on Scala 2.13 with DATAGEN_SEED=1716656294

**Describe the bug**

[2024-05-25T18:30:21.197Z] FAILED ../../src/main/python/conditionals_test.py::test_conditional_with_side_effects_case_when[String][DATAGEN_SEED=1716656294, TZ=UTC, INJECT_OOM] - py4j.protocol.Py4JJavaError: An error occurred while calling o467059.collectToPython.

[2024-05-25T18:30:21.197Z] : org.apache.spark.SparkException: Job aborted due to stage failure: Task 1 in stage 6178.0 failed 1 times, most recent failure: Lost task 1.0 in stage 6178.0 (TID 11385) (100.102.147.209 executor 0): com.nvidia.spark.rapids.jni.CastException: Error casting data on row 0: R-

<details><summary>Details...</summary>
<pre>
[2024-05-25T18:30:21.190Z] _____________ test_conditional_with_side_effects_case_when[String] _____________
[2024-05-25T18:30:21.190Z] [gw2] linux -- Python 3.9.19 /opt/conda/bin/python
[2024-05-25T18:30:21.190Z] 
[2024-05-25T18:30:21.190Z] data_gen = String
[2024-05-25T18:30:21.190Z] 
[2024-05-25T18:30:21.190Z]     @pytest.mark.parametrize('data_gen', [mk_str_gen('[0-9]{1,9}')], ids=idfn)
[2024-05-25T18:30:21.190Z]     @pytest.mark.skipif(not is_jvm_charset_utf8(), reason="regular expressions require UTF-8")
[2024-05-25T18:30:21.190Z]     def test_conditional_with_side_effects_case_when(data_gen):
[2024-05-25T18:30:21.190Z]         test_conf=copy_and_update(
[2024-05-25T18:30:21.190Z]             ansi_enabled_conf, {'spark.rapids.sql.regexp.enabled': True})
[2024-05-25T18:30:21.190Z] >       assert_gpu_and_cpu_are_equal_collect(
[2024-05-25T18:30:21.190Z]                 lambda spark : unary_op_df(spark, data_gen).selectExpr(
[2024-05-25T18:30:21.190Z]                     'CASE \
[2024-05-25T18:30:21.190Z]                     WHEN a RLIKE "^[0-9]{6}" THEN CAST(SUBSTR(a, 0, 6) AS INT) + 123 \
[2024-05-25T18:30:21.190Z]                     WHEN a RLIKE "^[0-9]{3}" THEN CAST(SUBSTR(a, 0, 3) AS INT) \
[2024-05-25T18:30:21.190Z]                     ELSE -1 END'),
[2024-05-25T18:30:21.190Z]                     conf = test_conf)
[2024-05-25T18:30:21.190Z] 
[2024-05-25T18:30:21.190Z] ../../src/main/python/conditionals_test.py:227: 
[2024-05-25T18:30:21.190Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[2024-05-25T18:30:21.190Z] ../../src/main/python/asserts.py:595: in assert_gpu_and_cpu_are_equal_collect
[2024-05-25T18:30:21.190Z]     _assert_gpu_and_cpu_are_equal(func, 'COLLECT', conf=conf, is_cpu_first=is_cpu_first, result_canonicalize_func_before_compare=result_canonicalize_func_before_compare)
[2024-05-25T18:30:21.190Z] ../../src/main/python/asserts.py:503: in _assert_gpu_and_cpu_are_equal
[2024-05-25T18:30:21.190Z]     from_gpu = run_on_gpu()
[2024-05-25T18:30:21.190Z] ../../src/main/python/asserts.py:496: in run_on_gpu
[2024-05-25T18:30:21.190Z]     from_gpu = with_gpu_session(bring_back, conf=conf)
[2024-05-25T18:30:21.190Z] ../../src/main/python/spark_session.py:164: in with_gpu_session
[2024-05-25T18:30:21.190Z]     return with_spark_session(func, conf=copy)
[2024-05-25T18:30:21.190Z] /opt/conda/lib/python3.9/contextlib.py:79: in inner
[2024-05-25T18:30:21.190Z]     return func(*args, **kwds)
[2024-05-25T18:30:21.190Z] ../../src/main/python/spark_session.py:131: in with_spark_session
[2024-05-25T18:30:21.190Z]     ret = func(_spark)
[2024-05-25T18:30:21.190Z] ../../src/main/python/asserts.py:205: in <lambda>
[2024-05-25T18:30:21.190Z]     bring_back = lambda spark: limit_func(spark).collect()
[2024-05-25T18:30:21.190Z] ../../../spark-3.5.0-bin-hadoop3-scala2.13/python/pyspark/sql/dataframe.py:1257: in collect
[2024-05-25T18:30:21.190Z]     sock_info = self._jdf.collectToPython()
[2024-05-25T18:30:21.190Z] /home/jenkins/agent/workspace/jenkins-rapids_integration-scala213-pre_release-github-83-3.5.0-cuda11/jars/spark-3.5.0-bin-hadoop3-scala2.13/python/lib/py4j-0.10.9.7-src.zip/py4j/java_gateway.py:1322: in __call__
[2024-05-25T18:30:21.190Z]     return_value = get_return_value(
[2024-05-25T18:30:21.190Z] ../../../spark-3.5.0-bin-hadoop3-scala2.13/python/pyspark/errors/exceptions/captured.py:179: in deco
[2024-05-25T18:30:21.190Z]     return f(*a, **kw)
[2024-05-25T18:30:21.190Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[2024-05-25T18:30:21.190Z] 
[2024-05-25T18:30:21.190Z] answer = 'xro467060'
[2024-05-25T18:30:21.190Z] gateway_client = <py4j.clientserver.JavaClient object at 0x7f7ab26214c0>
[2024-05-25T18:30:21.190Z] target_id = 'o467059', name = 'collectToPython'
[2024-05-25T18:30:21.190Z] 
[2024-05-25T18:30:21.190Z]     def get_return_value(answer, gateway_client, target_id=None, name=None):
[2024-05-25T18:30:21.190Z]         """Converts an answer received from the Java gateway into a Python object.
[2024-05-25T18:30:21.190Z]     
[2024-05-25T18:30:21.190Z]         For example, string representation of integers are converted to Python
[2024-05-25T18:30:21.190Z]         integer, string representation of objects are converted to JavaObject
[2024-05-25T18:30:21.190Z]         instances, etc.
[2024-05-25T18:30:21.190Z]     
[2024-05-25T18:30:21.190Z]         :param answer: the string returned by the Java gateway
[2024-05-25T18:30:21.190Z]         :param gateway_client: the gateway client used to communicate with the Java
[2024-05-25T18:30:21.190Z]             Gateway. Only necessary if the answer is a reference (e.g., object,
[2024-05-25T18:30:21.190Z]             list, map)
[2024-05-25T18:30:21.190Z]         :param target_id: the name of the object from which the answer comes from
[2024-05-25T18:30:21.190Z]             (e.g., *object1* in `object1.hello()`). Optional.
[2024-05-25T18:30:21.190Z]         :param name: the name of the member from which the answer comes from
[2024-05-25T18:30:21.190Z]             (e.g., *hello* in `object1.hello()`). Optional.
[2024-05-25T18:30:21.190Z]         """
[2024-05-25T18:30:21.190Z]         if is_error(answer)[0]:
[2024-05-25T18:30:21.191Z]             if len(answer) > 1:
[2024-05-25T18:30:21.191Z]                 type = answer[1]
[2024-05-25T18:30:21.191Z]                 value = OUTPUT_CONVERTER[type](answer[2:], gateway_client)
[2024-05-25T18:30:21.191Z]                 if answer[1] == REFERENCE_TYPE:
[2024-05-25T18:30:21.191Z] >                   raise Py4JJavaError(
[2024-05-25T18:30:21.191Z]                         "An error occurred while calling {0}{1}{2}.\n".
[2024-05-25T18:30:21.191Z]                         format(target_id, ".", name), value)
[2024-05-25T18:30:21.191Z] E                   py4j.protocol.Py4JJavaError: An error occurred while calling o467059.collectToPython.
[2024-05-25T18:30:21.191Z] E                   : org.apache.spark.SparkException: Job aborted due to stage failure: Task 1 in stage 6178.0 failed 1 times, most recent failure: Lost task 1.0 in stage 6178.0 (TID 11385) (100.102.147.209 executor 0): com.nvidia.spark.rapids.jni.CastException: Error casting data on row 0: R-
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.jni.CastStrings.toInteger(Native Method)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.jni.CastStrings.toInteger(CastStrings.java:50)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.jni.CastStrings.toInteger(CastStrings.java:37)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.GpuCast$.doCast(GpuCast.scala:548)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.GpuCast.doColumnar(GpuCast.scala:1903)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.GpuUnaryExpression.doItColumnar(GpuExpressions.scala:250)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.GpuUnaryExpression.$anonfun$columnarEval$1(GpuExpressions.scala:261)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.GpuUnaryExpression.columnarEval(GpuExpressions.scala:260)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$ReallyAGpuExpression.columnarEval(implicits.scala:34)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.GpuCaseWhen.$anonfun$filterEvaluateWhenThen$1(conditionalExpressions.scala:468)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.GpuCaseWhen.filterEvaluateWhenThen(conditionalExpressions.scala:467)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.GpuCaseWhen.$anonfun$columnarEvalWithSideEffects$4(conditionalExpressions.scala:407)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.GpuCaseWhen.$anonfun$columnarEvalWithSideEffects$4$adapted(conditionalExpressions.scala:400)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.GpuCaseWhen.$anonfun$columnarEvalWithSideEffects$3(conditionalExpressions.scala:400)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.GpuCaseWhen.$anonfun$columnarEvalWithSideEffects$3$adapted(conditionalExpressions.scala:396)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.GpuCaseWhen.$anonfun$columnarEvalWithSideEffects$2(conditionalExpressions.scala:396)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.GpuCaseWhen.$anonfun$columnarEvalWithSideEffects$2$adapted(conditionalExpressions.scala:393)
[2024-05-25T18:30:21.191Z] E                   	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:563)
[2024-05-25T18:30:21.191Z] E                   	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:561)
[2024-05-25T18:30:21.191Z] E                   	at scala.collection.AbstractIterable.foreach(Iterable.scala:926)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.GpuCaseWhen.$anonfun$columnarEvalWithSideEffects$1(conditionalExpressions.scala:393)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.GpuCaseWhen.columnarEvalWithSideEffects(conditionalExpressions.scala:390)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.GpuCaseWhen.columnarEval(conditionalExpressions.scala:360)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$ReallyAGpuExpression.columnarEval(implicits.scala:34)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.GpuAlias.columnarEval(namedExpressions.scala:110)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$ReallyAGpuExpression.columnarEval(implicits.scala:34)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.GpuProjectExec$.$anonfun$project$1(basicPhysicalOperators.scala:110)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$IterableMapsSafely.$anonfun$safeMap$1(implicits.scala:220)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$IterableMapsSafely.$anonfun$safeMap$1$adapted(implicits.scala:217)
[2024-05-25T18:30:21.191Z] E                   	at scala.collection.immutable.List.foreach(List.scala:333)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$IterableMapsSafely.safeMap(implicits.scala:217)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$AutoCloseableProducingSeq.safeMap(implicits.scala:331)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.GpuProjectExec$.project(basicPhysicalOperators.scala:110)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.GpuTieredProject.$anonfun$project$2(basicPhysicalOperators.scala:619)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.GpuTieredProject.recurse$2(basicPhysicalOperators.scala:618)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.GpuTieredProject.project(basicPhysicalOperators.scala:631)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.GpuTieredProject.$anonfun$projectWithRetrySingleBatchInternal$5(basicPhysicalOperators.scala:567)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$.withRestoreOnRetry(RmmRapidsRetryIterator.scala:272)
[2024-05-25T18:30:21.191Z] E                   	at com.nvidia.spark.rapids.GpuTieredProject.$anonfun$projectWithRetrySingleBatchInternal$4(basicPhysicalOperators.scala:567)
[2024-05-25T18:30:21.192Z] E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
[2024-05-25T18:30:21.192Z] E                   	at com.nvidia.spark.rapids.GpuTieredProject.$anonfun$projectWithRetrySingleBatchInternal$3(basicPhysicalOperators.scala:565)
[2024-05-25T18:30:21.192Z] E                   	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$NoInputSpliterator.next(RmmRapidsRetryIterator.scala:395)
[2024-05-25T18:30:21.192Z] E                   	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$RmmRapidsRetryIterator.next(RmmRapidsRetryIterator.scala:613)
[2024-05-25T18:30:21.192Z] E                   	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$RmmRapidsRetryAutoCloseableIterator.next(RmmRapidsRetryIterator.scala:517)
[2024-05-25T18:30:21.192Z] E                   	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$.drainSingleWithVerification(RmmRapidsRetryIterator.scala:291)
[2024-05-25T18:30:21.192Z] E                   	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$.withRetryNoSplit(RmmRapidsRetryIterator.scala:185)
[2024-05-25T18:30:21.192Z] E                   	at com.nvidia.spark.rapids.GpuTieredProject.$anonfun$projectWithRetrySingleBatchInternal$1(basicPhysicalOperators.scala:565)
[2024-05-25T18:30:21.192Z] E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:39)
[2024-05-25T18:30:21.192Z] E                   	at com.nvidia.spark.rapids.GpuTieredProject.projectWithRetrySingleBatchInternal(basicPhysicalOperators.scala:562)
[2024-05-25T18:30:21.192Z] E                   	at com.nvidia.spark.rapids.GpuTieredProject.projectAndCloseWithRetrySingleBatch(basicPhysicalOperators.scala:601)
[2024-05-25T18:30:21.192Z] E                   	at com.nvidia.spark.rapids.GpuProjectExec.$anonfun$internalDoExecuteColumnar$2(basicPhysicalOperators.scala:384)
[2024-05-25T18:30:21.192Z] E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
[2024-05-25T18:30:21.192Z] E                   	at com.nvidia.spark.rapids.GpuProjectExec.$anonfun$internalDoExecuteColumnar$1(basicPhysicalOperators.scala:380)
[2024-05-25T18:30:21.192Z] E                   	at scala.collection.Iterator$$anon$9.next(Iterator.scala:577)
[2024-05-25T18:30:21.192Z] E                   	at com.nvidia.spark.rapids.ColumnarToRowIterator.$anonfun$fetchNextBatch$3(GpuColumnarToRowExec.scala:290)
[2024-05-25T18:30:21.192Z] E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
[2024-05-25T18:30:21.192Z] E                   	at com.nvidia.spark.rapids.ColumnarToRowIterator.fetchNextBatch(GpuColumnarToRowExec.scala:287)
[2024-05-25T18:30:21.192Z] E                   	at com.nvidia.spark.rapids.ColumnarToRowIterator.loadNextBatch(GpuColumnarToRowExec.scala:257)
[2024-05-25T18:30:21.192Z] E                   	at com.nvidia.spark.rapids.ColumnarToRowIterator.hasNext(GpuColumnarToRowExec.scala:304)
[2024-05-25T18:30:21.192Z] E                   	at scala.collection.Iterator$$anon$9.hasNext(Iterator.scala:576)
[2024-05-25T18:30:21.192Z] E                   	at org.apache.spark.sql.execution.SparkPlan.$anonfun$getByteArrayRdd$1(SparkPlan.scala:388)
[2024-05-25T18:30:21.192Z] E                   	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2(RDD.scala:890)
[2024-05-25T18:30:21.192Z] E                   	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2$adapted(RDD.scala:890)
[2024-05-25T18:30:21.192Z] E                   	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
[2024-05-25T18:30:21.192Z] E                   	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
[2024-05-25T18:30:21.192Z] E                   	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
[2024-05-25T18:30:21.192Z] E                   	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:93)
[2024-05-25T18:30:21.192Z] E                   	at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:161)
[2024-05-25T18:30:21.192Z] E                   	at org.apache.spark.scheduler.Task.run(Task.scala:141)
[2024-05-25T18:30:21.192Z] E                   	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$4(Executor.scala:620)
[2024-05-25T18:30:21.192Z] E                   	at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally(SparkErrorUtils.scala:64)
[2024-05-25T18:30:21.192Z] E                   	at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally$(SparkErrorUtils.scala:61)
[2024-05-25T18:30:21.192Z] E                   	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:94)
[2024-05-25T18:30:21.192Z] E                   	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:623)
[2024-05-25T18:30:21.192Z] E                   	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
[2024-05-25T18:30:21.192Z] E                   	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
[2024-05-25T18:30:21.192Z] E                   	at java.base/java.lang.Thread.run(Thread.java:840)
[2024-05-25T18:30:21.192Z] E                   	Suppressed: com.nvidia.spark.rapids.jni.GpuRetryOOM: injected RetryOOM
[2024-05-25T18:30:21.192Z] E                   		at ai.rapids.cudf.Scalar.makeStringScalar(Native Method)
[2024-05-25T18:30:21.192Z] E                   		at ai.rapids.cudf.Scalar.fromUTF8String(Scalar.java:366)
[2024-05-25T18:30:21.192Z] E                   		at ai.rapids.cudf.Scalar.fromString(Scalar.java:354)
[2024-05-25T18:30:21.192Z] E                   		at com.nvidia.spark.rapids.GpuScalar$.from(literals.scala:328)
[2024-05-25T18:30:21.192Z] E                   		at com.nvidia.spark.rapids.GpuScalar.getBase(literals.scala:525)
[2024-05-25T18:30:21.192Z] E                   		at org.apache.spark.sql.rapids.GpuLiteralRangePattern.doColumnar(stringFunctions.scala:1138)
[2024-05-25T18:30:21.192Z] E                   		at com.nvidia.spark.rapids.GpuBinaryExpression.$anonfun$columnarEval$3(GpuExpressions.scala:318)
[2024-05-25T18:30:21.192Z] E                   		at com.nvidia.spark.rapids.Arm$.withResourceIfAllowed(Arm.scala:84)
[2024-05-25T18:30:21.192Z] E                   		at com.nvidia.spark.rapids.GpuBinaryExpression.$anonfun$columnarEval$2(GpuExpressions.scala:311)
[2024-05-25T18:30:21.192Z] E                   		at com.nvidia.spark.rapids.Arm$.withResourceIfAllowed(Arm.scala:84)
[2024-05-25T18:30:21.192Z] E                   		at com.nvidia.spark.rapids.GpuBinaryExpression.columnarEval(GpuExpressions.scala:310)
[2024-05-25T18:30:21.192Z] E                   		at com.nvidia.spark.rapids.GpuBinaryExpression.columnarEval$(GpuExpressions.scala:309)
[2024-05-25T18:30:21.192Z] E                   		at org.apache.spark.sql.rapids.GpuLiteralRangePattern.columnarEval(stringFunctions.scala:1129)
[2024-05-25T18:30:21.192Z] E                   		at com.nvidia.spark.rapids.RapidsPluginImplicits$ReallyAGpuExpression.columnarEval(implicits.scala:34)
[2024-05-25T18:30:21.192Z] E                   		... 64 more
[2024-05-25T18:30:21.192Z] E                   
[2024-05-25T18:30:21.192Z] E                   Driver stacktrace:
[2024-05-25T18:30:21.192Z] E                   	at org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2844)
[2024-05-25T18:30:21.192Z] E                   	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2780)
[2024-05-25T18:30:21.192Z] E                   	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2779)
[2024-05-25T18:30:21.192Z] E                   	at scala.collection.immutable.List.foreach(List.scala:333)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2779)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:1242)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:1242)
[2024-05-25T18:30:21.193Z] E                   	at scala.Option.foreach(Option.scala:437)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:1242)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:3048)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2982)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2971)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:984)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2398)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2419)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2438)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2463)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.rdd.RDD.$anonfun$collect$1(RDD.scala:1046)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.rdd.RDD.withScope(RDD.scala:407)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.rdd.RDD.collect(RDD.scala:1045)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.sql.execution.SparkPlan.executeCollect(SparkPlan.scala:448)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.sql.Dataset.$anonfun$collectToPython$1(Dataset.scala:4160)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.sql.Dataset.$anonfun$withAction$2(Dataset.scala:4334)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.sql.execution.QueryExecution$.withInternalError(QueryExecution.scala:546)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.sql.Dataset.$anonfun$withAction$1(Dataset.scala:4332)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$6(SQLExecution.scala:125)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:201)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:108)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:900)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:66)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.sql.Dataset.withAction(Dataset.scala:4332)
[2024-05-25T18:30:21.193Z] E                   	at org.apache.spark.sql.Dataset.collectToPython(Dataset.scala:4157)
[2024-05-25T18:30:21.193Z] E                   	at jdk.internal.reflect.GeneratedMethodAccessor68.invoke(Unknown Source)
[2024-05-25T18:30:21.193Z] E                   	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[2024-05-25T18:30:21.193Z] E                   	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
[2024-05-25T18:30:21.193Z] E                   	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
[2024-05-25T18:30:21.193Z] E                   	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:374)
[2024-05-25T18:30:21.193Z] E                   	at py4j.Gateway.invoke(Gateway.java:282)
[2024-05-25T18:30:21.193Z] E                   	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
[2024-05-25T18:30:21.193Z] E                   	at py4j.commands.CallCommand.execute(CallCommand.java:79)
[2024-05-25T18:30:21.193Z] E                   	at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)
[2024-05-25T18:30:21.193Z] E                   	at py4j.ClientServerConnection.run(ClientServerConnection.java:106)
[2024-05-25T18:30:21.193Z] E                   	at java.base/java.lang.Thread.run(Thread.java:840)
[2024-05-25T18:30:21.193Z] E                   Caused by: com.nvidia.spark.rapids.jni.CastException: Error casting data on row 0: R-
[2024-05-25T18:30:21.193Z] E                   	at com.nvidia.spark.rapids.jni.CastStrings.toInteger(Native Method)
[2024-05-25T18:30:21.193Z] E                   	at com.nvidia.spark.rapids.jni.CastStrings.toInteger(CastStrings.java:50)
[2024-05-25T18:30:21.193Z] E                   	at com.nvidia.spark.rapids.jni.CastStrings.toInteger(CastStrings.java:37)
[2024-05-25T18:30:21.193Z] E                   	at com.nvidia.spark.rapids.GpuCast$.doCast(GpuCast.scala:548)
[2024-05-25T18:30:21.193Z] E                   	at com.nvidia.spark.rapids.GpuCast.doColumnar(GpuCast.scala:1903)
[2024-05-25T18:30:21.193Z] E                   	at com.nvidia.spark.rapids.GpuUnaryExpression.doItColumnar(GpuExpressions.scala:250)
[2024-05-25T18:30:21.193Z] E                   	at com.nvidia.spark.rapids.GpuUnaryExpression.$anonfun$columnarEval$1(GpuExpressions.scala:261)
[2024-05-25T18:30:21.193Z] E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
[2024-05-25T18:30:21.193Z] E                   	at com.nvidia.spark.rapids.GpuUnaryExpression.columnarEval(GpuExpressions.scala:260)
[2024-05-25T18:30:21.193Z] E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$ReallyAGpuExpression.columnarEval(implicits.scala:34)
[2024-05-25T18:30:21.193Z] E                   	at com.nvidia.spark.rapids.GpuCaseWhen.$anonfun$filterEvaluateWhenThen$1(conditionalExpressions.scala:468)
[2024-05-25T18:30:21.193Z] E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
[2024-05-25T18:30:21.193Z] E                   	at com.nvidia.spark.rapids.GpuCaseWhen.filterEvaluateWhenThen(conditionalExpressions.scala:467)
[2024-05-25T18:30:21.193Z] E                   	at com.nvidia.spark.rapids.GpuCaseWhen.$anonfun$columnarEvalWithSideEffects$4(conditionalExpressions.scala:407)
[2024-05-25T18:30:21.193Z] E                   	at com.nvidia.spark.rapids.GpuCaseWhen.$anonfun$columnarEvalWithSideEffects$4$adapted(conditionalExpressions.scala:400)
[2024-05-25T18:30:21.193Z] E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
[2024-05-25T18:30:21.193Z] E                   	at com.nvidia.spark.rapids.GpuCaseWhen.$anonfun$columnarEvalWithSideEffects$3(conditionalExpressions.scala:400)
[2024-05-25T18:30:21.193Z] E                   	at com.nvidia.spark.rapids.GpuCaseWhen.$anonfun$columnarEvalWithSideEffects$3$adapted(conditionalExpressions.scala:396)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.GpuCaseWhen.$anonfun$columnarEvalWithSideEffects$2(conditionalExpressions.scala:396)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.GpuCaseWhen.$anonfun$columnarEvalWithSideEffects$2$adapted(conditionalExpressions.scala:393)
[2024-05-25T18:30:21.194Z] E                   	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:563)
[2024-05-25T18:30:21.194Z] E                   	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:561)
[2024-05-25T18:30:21.194Z] E                   	at scala.collection.AbstractIterable.foreach(Iterable.scala:926)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.GpuCaseWhen.$anonfun$columnarEvalWithSideEffects$1(conditionalExpressions.scala:393)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.GpuCaseWhen.columnarEvalWithSideEffects(conditionalExpressions.scala:390)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.GpuCaseWhen.columnarEval(conditionalExpressions.scala:360)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$ReallyAGpuExpression.columnarEval(implicits.scala:34)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.GpuAlias.columnarEval(namedExpressions.scala:110)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$ReallyAGpuExpression.columnarEval(implicits.scala:34)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.GpuProjectExec$.$anonfun$project$1(basicPhysicalOperators.scala:110)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$IterableMapsSafely.$anonfun$safeMap$1(implicits.scala:220)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$IterableMapsSafely.$anonfun$safeMap$1$adapted(implicits.scala:217)
[2024-05-25T18:30:21.194Z] E                   	at scala.collection.immutable.List.foreach(List.scala:333)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$IterableMapsSafely.safeMap(implicits.scala:217)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.RapidsPluginImplicits$AutoCloseableProducingSeq.safeMap(implicits.scala:331)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.GpuProjectExec$.project(basicPhysicalOperators.scala:110)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.GpuTieredProject.$anonfun$project$2(basicPhysicalOperators.scala:619)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.GpuTieredProject.recurse$2(basicPhysicalOperators.scala:618)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.GpuTieredProject.project(basicPhysicalOperators.scala:631)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.GpuTieredProject.$anonfun$projectWithRetrySingleBatchInternal$5(basicPhysicalOperators.scala:567)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$.withRestoreOnRetry(RmmRapidsRetryIterator.scala:272)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.GpuTieredProject.$anonfun$projectWithRetrySingleBatchInternal$4(basicPhysicalOperators.scala:567)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.GpuTieredProject.$anonfun$projectWithRetrySingleBatchInternal$3(basicPhysicalOperators.scala:565)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$NoInputSpliterator.next(RmmRapidsRetryIterator.scala:395)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$RmmRapidsRetryIterator.next(RmmRapidsRetryIterator.scala:613)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$RmmRapidsRetryAutoCloseableIterator.next(RmmRapidsRetryIterator.scala:517)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$.drainSingleWithVerification(RmmRapidsRetryIterator.scala:291)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$.withRetryNoSplit(RmmRapidsRetryIterator.scala:185)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.GpuTieredProject.$anonfun$projectWithRetrySingleBatchInternal$1(basicPhysicalOperators.scala:565)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:39)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.GpuTieredProject.projectWithRetrySingleBatchInternal(basicPhysicalOperators.scala:562)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.GpuTieredProject.projectAndCloseWithRetrySingleBatch(basicPhysicalOperators.scala:601)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.GpuProjectExec.$anonfun$internalDoExecuteColumnar$2(basicPhysicalOperators.scala:384)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.GpuProjectExec.$anonfun$internalDoExecuteColumnar$1(basicPhysicalOperators.scala:380)
[2024-05-25T18:30:21.194Z] E                   	at scala.collection.Iterator$$anon$9.next(Iterator.scala:577)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.ColumnarToRowIterator.$anonfun$fetchNextBatch$3(GpuColumnarToRowExec.scala:290)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.ColumnarToRowIterator.fetchNextBatch(GpuColumnarToRowExec.scala:287)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.ColumnarToRowIterator.loadNextBatch(GpuColumnarToRowExec.scala:257)
[2024-05-25T18:30:21.194Z] E                   	at com.nvidia.spark.rapids.ColumnarToRowIterator.hasNext(GpuColumnarToRowExec.scala:304)
[2024-05-25T18:30:21.194Z] E                   	at scala.collection.Iterator$$anon$9.hasNext(Iterator.scala:576)
[2024-05-25T18:30:21.194Z] E                   	at org.apache.spark.sql.execution.SparkPlan.$anonfun$getByteArrayRdd$1(SparkPlan.scala:388)
[2024-05-25T18:30:21.194Z] E                   	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2(RDD.scala:890)
[2024-05-25T18:30:21.194Z] E                   	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2$adapted(RDD.scala:890)
[2024-05-25T18:30:21.194Z] E                   	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
[2024-05-25T18:30:21.194Z] E                   	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
[2024-05-25T18:30:21.194Z] E                   	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
[2024-05-25T18:30:21.194Z] E                   	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:93)
[2024-05-25T18:30:21.194Z] E                   	at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:161)
[2024-05-25T18:30:21.194Z] E                   	at org.apache.spark.scheduler.Task.run(Task.scala:141)
[2024-05-25T18:30:21.195Z] E                   	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$4(Executor.scala:620)
[2024-05-25T18:30:21.195Z] E                   	at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally(SparkErrorUtils.scala:64)
[2024-05-25T18:30:21.195Z] E                   	at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally$(SparkErrorUtils.scala:61)
[2024-05-25T18:30:21.195Z] E                   	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:94)
[2024-05-25T18:30:21.195Z] E                   	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:623)
[2024-05-25T18:30:21.195Z] E                   	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
[2024-05-25T18:30:21.195Z] E                   	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
[2024-05-25T18:30:21.195Z] E                   	... 1 more
[2024-05-25T18:30:21.195Z] E                   	Suppressed: com.nvidia.spark.rapids.jni.GpuRetryOOM: injected RetryOOM
[2024-05-25T18:30:21.195Z] E                   		at ai.rapids.cudf.Scalar.makeStringScalar(Native Method)
[2024-05-25T18:30:21.195Z] E                   		at ai.rapids.cudf.Scalar.fromUTF8String(Scalar.java:366)
[2024-05-25T18:30:21.195Z] E                   		at ai.rapids.cudf.Scalar.fromString(Scalar.java:354)
[2024-05-25T18:30:21.195Z] E                   		at com.nvidia.spark.rapids.GpuScalar$.from(literals.scala:328)
[2024-05-25T18:30:21.195Z] E                   		at com.nvidia.spark.rapids.GpuScalar.getBase(literals.scala:525)
[2024-05-25T18:30:21.195Z] E                   		at org.apache.spark.sql.rapids.GpuLiteralRangePattern.doColumnar(stringFunctions.scala:1138)
[2024-05-25T18:30:21.195Z] E                   		at com.nvidia.spark.rapids.GpuBinaryExpression.$anonfun$columnarEval$3(GpuExpressions.scala:318)
[2024-05-25T18:30:21.195Z] E                   		at com.nvidia.spark.rapids.Arm$.withResourceIfAllowed(Arm.scala:84)
[2024-05-25T18:30:21.195Z] E                   		at com.nvidia.spark.rapids.GpuBinaryExpression.$anonfun$columnarEval$2(GpuExpressions.scala:311)
[2024-05-25T18:30:21.195Z] E                   		at com.nvidia.spark.rapids.Arm$.withResourceIfAllowed(Arm.scala:84)
[2024-05-25T18:30:21.195Z] E                   		at com.nvidia.spark.rapids.GpuBinaryExpression.columnarEval(GpuExpressions.scala:310)
[2024-05-25T18:30:21.195Z] E                   		at com.nvidia.spark.rapids.GpuBinaryExpression.columnarEval$(GpuExpressions.scala:309)
[2024-05-25T18:30:21.195Z] E                   		at org.apache.spark.sql.rapids.GpuLiteralRangePattern.columnarEval(stringFunctions.scala:1129)
[2024-05-25T18:30:21.195Z] E                   		at com.nvidia.spark.rapids.RapidsPluginImplicits$ReallyAGpuExpression.columnarEval(implicits.scala:34)
[2024-05-25T18:30:21.195Z] E                   		... 64 more
[2024-05-25T18:30:21.195Z] 
[2024-05-25T18:30:21.195Z] /home/jenkins/agent/workspace/jenkins-rapids_integration-scala213-pre_release-github-83-3.5.0-cuda11/jars/spark-3.5.0-bin-hadoop3-scala2.13/python/lib/py4j-0.10.9.7-src.zip/py4j/protocol.py:326: Py4JJavaError
</pre>
</details>

**Steps/Code to reproduce bug**
DATAGEN_SEED=DATAGEN_SEED=1716656294 on 24.06

**Expected behavior**
Test should pass

**Environment details (please complete the following information)**
Scala 2.13 integration test on x86

**Additional context**

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala
@@ -2125,11 +2125,14 @@ object RegexRewrite {
    */
   def matchSimplePattern(ast: RegexAST): RegexOptimizationType = {
     ast.children() match {
-      case (RegexChar('^') | RegexEscaped('A')) :: ast 
-          if isliteralString(stripTailingWildcards(ast)) => {
-        // ^literal.* => startsWith literal
-        RegexOptimizationType.StartsWith(RegexCharsToString(stripTailingWildcards(ast)))
-      }
+      case (RegexChar('^') | RegexEscaped('A')) :: ast =>
+        val noWildCardsAst = stripTailingWildcards(ast)
+        if (isliteralString(noWildCardsAst)) {
+          // ^literal.* => startsWith literal
+          RegexOptimizationType.StartsWith(RegexCharsToString(stripTailingWildcards(ast)))
+        } else {
+          RegexOptimizationType.NoOptimization
+        }
       case astLs => {
         val noStartsWithAst = stripTailingWildcards(stripLeadingWildcards(astLs))
         val prefixRangeInfo = getPrefixRangePattern(noStartsWithAst)

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala
@@ -17,7 +17,6 @@ package com.nvidia.spark.rapids
 
 import java.sql.SQLException
 
-import scala.collection
 import scala.collection.mutable.ListBuffer
 
 import com.nvidia.spark.rapids.GpuOverrides.regexMetaChars
@@ -73,7 +72,7 @@ class RegexParser(pattern: String) {
     sequence
   }
 
-  def parseReplacementBase(): RegexAST = {
+  private def parseReplacementBase(): RegexAST = {
       consume() match {
         case '\\' =>
           parseBackrefOrEscaped()
@@ -782,6 +781,7 @@ class CudfRegexTranspiler(mode: RegexMode) {
     }
   }
 
+  @scala.annotation.tailrec
   private def isRepetition(e: RegexAST, checkZeroLength: Boolean): Boolean = {
     e match {
       case RegexRepetition(_, _) if !checkZeroLength => true
@@ -1648,6 +1648,7 @@ class CudfRegexTranspiler(mode: RegexMode) {
     }
   }
 
+  @scala.annotation.tailrec
   private def isEntirely(regex: RegexAST, f: RegexAST => Boolean): Boolean = {
     regex match {
       case RegexSequence(parts) if parts.nonEmpty =>
@@ -1672,6 +1673,7 @@ class CudfRegexTranspiler(mode: RegexMode) {
     })
   }
 
+  @scala.annotation.tailrec
   private def beginsWith(regex: RegexAST, f: RegexAST => Boolean): Boolean = {
     regex match {
       case RegexSequence(parts) if parts.nonEmpty =>
@@ -1687,6 +1689,7 @@ class CudfRegexTranspiler(mode: RegexMode) {
 
   }
 
+  @scala.annotation.tailrec
   private def endsWith(regex: RegexAST, f: RegexAST => Boolean): Boolean = {
     regex match {
       case RegexSequence(parts) if parts.nonEmpty =>
@@ -1760,7 +1763,7 @@ sealed case class RegexSequence(parts: ListBuffer[RegexAST]) extends RegexAST {
 }
 
 sealed case class RegexGroup(capture: Boolean, term: RegexAST,
-    val lookahead: Option[RegexLookahead])
+    lookahead: Option[RegexLookahead])
     extends RegexAST {
   def this(capture: Boolean, term: RegexAST) = {
     this(capture, term, None)
@@ -2028,6 +2031,7 @@ object RegexOptimizationType {
 
 object RegexRewrite {
 
+  @scala.annotation.tailrec
   private def removeBrackets(astLs: collection.Seq[RegexAST]): collection.Seq[RegexAST] = {
     astLs match {
       case collection.Seq(RegexGroup(_, term, None)) => removeBrackets(term.children())
@@ -2044,7 +2048,7 @@ object RegexRewrite {
    */
   private def getPrefixRangePattern(astLs: collection.Seq[RegexAST]): 
       Option[(String, Int, Int, Int)] = {
-    val haveLiteralPrefix = isliteralString(astLs.dropRight(1))
+    val haveLiteralPrefix = isLiteralString(astLs.dropRight(1))
     val endsWithRange = astLs.lastOption match {
       case Some(RegexRepetition(
           RegexCharacterClass(false, ListBuffer(RegexCharacterRange(a,b))), 
@@ -2080,9 +2084,9 @@ object RegexRewrite {
     }
   }
 
-  private def isliteralString(astLs: collection.Seq[RegexAST]): Boolean = {
+  private def isLiteralString(astLs: collection.Seq[RegexAST]): Boolean = {
     removeBrackets(astLs).forall {
-      case RegexChar(ch) if !regexMetaChars.contains(ch) => true
+      case RegexChar(ch) => !regexMetaChars.contains(ch)
       case _ => false
     }
   }
@@ -2120,18 +2124,25 @@ object RegexRewrite {
    * Matches the given regex ast to a regex optimization type for regex rewrite
    * optimization.
    *
-   * @param ast The Abstract Syntax Tree parsed from a regex pattern.
+   * @param ast unparsed children of the Abstract Syntax Tree parsed from a regex pattern.
    * @return The `RegexOptimizationType` for the given pattern.
    */
-  def matchSimplePattern(ast: RegexAST): RegexOptimizationType = {
-    ast.children() match {
-      case (RegexChar('^') | RegexEscaped('A')) :: ast =>
-        val noWildCardsAst = stripTailingWildcards(ast)
-        if (isliteralString(noWildCardsAst)) {
+  @scala.annotation.tailrec
+  def matchSimplePattern(ast: Seq[RegexAST]): RegexOptimizationType = {
+    ast match {
+      case (RegexChar('^') | RegexEscaped('A')) :: astTail =>
+        val noTrailingWildCards = stripTailingWildcards(astTail)
+        if (isLiteralString(noTrailingWildCards)) {
           // ^literal.* => startsWith literal
-          RegexOptimizationType.StartsWith(RegexCharsToString(stripTailingWildcards(ast)))
+          RegexOptimizationType.StartsWith(RegexCharsToString(noTrailingWildCards))
         } else {
-          RegexOptimizationType.NoOptimization
+          val noWildCards = stripLeadingWildcards(noTrailingWildCards)
+          if (noWildCards.length == noTrailingWildCards.length) {
+            // TODO startsWith with PrefIxRange
+            RegexOptimizationType.NoOptimization
+          } else {
+            matchSimplePattern(astTail)
+          }
         }
       case astLs => {
         val noStartsWithAst = stripTailingWildcards(stripLeadingWildcards(astLs))
@@ -2140,7 +2151,7 @@ object RegexRewrite {
           val (prefix, length, start, end) = prefixRangeInfo.get
           // (literal[a-b]{x,y}) => prefix range pattern
           RegexOptimizationType.PrefixRange(prefix, length, start, end)
-        } else if (isliteralString(noStartsWithAst)) {
+        } else if (isLiteralString(noStartsWithAst)) {
           // literal.* or (literal).* => contains literal
           RegexOptimizationType.Contains(RegexCharsToString(noStartsWithAst))
         } else {

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala
@@ -1073,7 +1073,7 @@ class GpuRLikeMeta(
             val originalPattern = str.toString
             val regexAst = new RegexParser(originalPattern).parse()
             if (conf.isRlikeRegexRewriteEnabled) {
-              rewriteOptimizationType = RegexRewrite.matchSimplePattern(regexAst)
+              rewriteOptimizationType = RegexRewrite.matchSimplePattern(regexAst.children())
             }
             val (transpiledAST, _) = new CudfRegexTranspiler(RegexFindMode)
                 .getTranspiledAST(regexAst, None, None)

--- a/tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionRewriteSuite.scala
+++ b/tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionRewriteSuite.scala
@@ -23,7 +23,7 @@ class RegularExpressionRewriteSuite extends AnyFunSuite {
       Unit = {
     val results = patterns.map { pattern =>
       val ast = new RegexParser(pattern).parse()
-      RegexRewrite.matchSimplePattern(ast)
+      RegexRewrite.matchSimplePattern(ast.children())
     }
     assert(results == excepted)
   }
@@ -53,12 +53,23 @@ class RegularExpressionRewriteSuite extends AnyFunSuite {
       "(.*)abc[0-9a-z]{1,3}(.*)",
       "(.*)abc[0-9]{2}.*",
       "^abc[0-9]{1,3}",
-      "火花急流[\u4e00-\u9fa5]{1}")
-    val excepted = Seq(PrefixRange("abc", 1, 48, 57),
-      NoOptimization,
-      PrefixRange("abc", 2, 48, 57),
+      "火花急流[\u4e00-\u9fa5]{1}",
+      "^[0-9]{6}",
+      "^[0-9]{3,10}",
+      "^.*[0-9]{6}",
+      "^(.*)[0-9]{3,10}"
+    )
+    val excepted = Seq(
       PrefixRange("abc", 1, 48, 57),
-      PrefixRange("火花急流", 1, 19968, 40869))
+      NoOptimization, // prefix followed by a multi-range not supported
+      PrefixRange("abc", 2, 48, 57),
+      NoOptimization, // starts with PrefixRange not supported
+      PrefixRange("火花急流", 1, 19968, 40869),
+      NoOptimization, // starts with PrefixRange not supported
+      NoOptimization, // starts with PrefixRange not supported
+      PrefixRange("", 6, 48, 57),
+      PrefixRange("", 3, 48, 57)
+    )
     verifyRewritePattern(patterns, excepted)
   }
 }

